lib/pp_diff.mli

```diff
@@ -0,0 +1,46 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(** Computes the diff between two Pp.t structures and return
+versions of each with diffs highlighted as (old, new) *)
+val diff_pp : Pp.t -> Pp.t -> Pp.t * Pp.t
+
+(** Computes the diff between two Pp.t structures and return
+a highlighted Pp.t.  If [show_removed] is true, show separate lines for
+removals and additions, otherwise only show additions *)
+val diff_pp_combined : ?show_removed:bool -> Pp.t -> Pp.t -> Pp.t
+
+(** Raised if the diff fails, e.g. the input can't be lexed *)
+exception Diff_Failure of string
+
+(* for dependency injection to allow calling the lexer *)
+val tokenize_string : (string -> string list) ref
+
```

Comment:
Make this an optional parameter of the diff function.

Reply:
Ah, do you recall the discussion we had about this earlier in this PR?  Search for "I have a problem creating pp_diff.ml in lib, since pp_diff () uses the lexer, a part of parsing, which comes after lib in the linking." above and several subsequent comments.

Making tokenize_string a parameter here just moves the forward-reference problem to callers of diff from modules before the `printing` module.  That seems like something we will want to do in the not-distant future.  Addressing the forward-reference in Pp solves the problem for all users.  I agree, dependency injection is not the simplest technique but sometimes it is invaluable.

As I said earlier "who wants to write another tokenizer that will be, let's face it, inferior?"

What say you?

Reply:
The problem of the reference is that it doesn't make explicit who's using it so functional unit testing becomes hard. I do prefer if you add a `?tokenizer:(string -> string list)` parameter to the places that are needed.

Reply:
Hard in what way?

It appears that the assignment to `Pp_diff.tokenize_string` is executed even if there's no function call to any method in the printing module or an `open Proof_diffs`.  (However, I didn't find anything in the OCaml doc explicitly specifying this behavior.)

I created the following unit test which I ran by itself.  It did the diff on a token basis, indicating that the reference was set.  If it used the default string_tokenize, it would have done diffs on a character by character basis.  So it appears no setup would be required to use diff_str in a functional unit test.

```
let _ = Pp_diff.tokenize_string := tokenize_string
```

```
open OUnit
open Utest
open Pp_diff

let tests = ref []
let add_test name test = tests := (mk_test name (TestCase test)) :: !tests

let t () =
  let expected : diff_list =
    [ `Removed (0, "abc"); `Added (0, "xabcy")] in
  let diffs = diff_str "abc" "xabcy   " in

  assert_equal ~msg:"real lexer" ~printer:string_of_diffs expected diffs
let _ = add_test "diff_str uses real lexer" t

let _ = run_tests __FILE__ (List.rev !tests)
```

The dependency injection makes this simple to use in all modules from "lib" on.  Without it, any call to `diff_str` in modules before "parsing" (if we move tokenize_string, or "printing" if we don't) will be markedly inferior.  And it seems likely folks will want to do diffs from some of those modules in the near future, which will re-raise the issue.

```
let core_src_dirs = [
  "config";
  "dev";
  "lib"; <- HOME OF pp_diff.ml, which needs tokenize_string
  "clib";
  "kernel";
  "library";
  "engine";
  "pretyping";
  "interp";
  "parsing"; <- HOME OF the lexer
  "proofs";
  "tactics";
  "toplevel";
  "printing"; <- CURRENT HOME OF tokenize_string (could be moved into "parsing" module
  "grammar";
  "ide";
  "stm";
  "vernac";
]
```

Somewhat off-point: For functional unit testing, I expect we will want to create some standard initialization routines that are shared among lots of tests.  For example, we may want to have a utility function that accepts a string, processes it as if the user typed it, then return the proof state for examination.  This probably requires some explicit setup beforehand, such as reading some of the theory files--which could be a standard initialization routine.  (On point: such standard initialization routines could set up tokenize_string although it appears that's not necessary.)

Take it from Humphrey Bogart:
"Rick Blaine: If you don't ... you'll regret it. Maybe not today, maybe not tomorrow, but soon and for the rest of your life."

Reply:
The same definition of functional unit testing rules out the reliance on global variables. I am not sure I understand what you mean in the rest of the comment, but I am 100% positive that the lexing function should be turned into a parameter [a quite simple change]